lib/c-api/src/wasm_c_api/instance.rs

```diff
@@ -32,7 +31,35 @@ pub unsafe extern "C" fn wasm_instance_new(
         .cloned()
         .collect();
 
-    let instance = Arc::new(c_try!(Instance::new(wasm_module, &resolver)));
+    let instance = match Instance::new(wasm_module, &resolver) {
+        Ok(instance) => Arc::new(instance),
+
+        Err(InstantiationError::Link(link_error)) => {
+            crate::error::update_last_error(link_error);
+
+            return None;
+        }
+
+        Err(InstantiationError::Start(runtime_error)) => {
+            let pointer = {
+                let trap: Box<wasm_trap_t> = Box::new(runtime_error.into());
+                let mut traps: Vec<*mut wasm_trap_t> = Vec::with_capacity(1);
+                traps.push(Box::into_raw(trap));
+
+                debug_assert!(traps.len() == traps.capacity());
+
+                let pointer = traps.as_mut_ptr();
+                mem::forget(traps);
```

Comment:
This code seems reasonable to me, ideally we could encapsulate more of it but I don't know if that's practical...

But without encapsulation, outside of the context of this PR diff, it's not obvious that this is a memory leak (which I believe it is)... we should leave a comment about it or try to fix it in this PR

Reply:
We leak it because we don't own it. The caller owns it.